### ISSUE TITLE
Preserve YMM/ZMM registers for runtime jithelpers

### ISSUE THREAD

**Issue by tajila** [eclipse-openj9/openj9#13004]
Continuing the work done here for https://github.com/eclipse-openj9/openj9/pull/12611 XMM registers.

JIT will be adding support for YMM/ZMM registers, we will need to be saving these between jit runtime helpers.

**Comment by tajila** [comment]
@BradleyWood can you please post resources to the linux and windows calling conventions for YMM and ZMM registers.

**Comment by tajila** [comment]
@BradleyWood Also, how does the JIT detect if YMM/ZMM registers are available on the host?

**Comment by gacholio** [comment]
We also need to know the instructions required to load/save the new registers, and required alignment of the memory being accessed.

**Comment by tajila** [comment]
@dnakamura please take a look at this

**Comment by gacholio** [comment]
If there are alignment restrictions for the new registers, we will need to forcibly align both the native and java stacks.

The native stack *may* already be aligned, the java stack is almost certainly not.

**Comment by gacholio** [comment]
Slight correction: the java stack only needs to be aligned if the JIT ever intends to spill one of the new registers (there are no preserved FPRs in the JIT private linkage, so it's possible to avoid this situation entirely).

**Comment by BradleyWood** [comment]
### Calling Conventions
https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-160&viewFallbackFrom=vs-2017
https://github.com/hjl-tools/x86-psABI/wiki/x86-64-psABI-1.0.pdf

I believe you can check for ymm/zmm register support by checking for OMR_FEATURE_X86_AVX2 and OMR_FEATURE_X86_AVX512F, respectively. 

You will want to use VMOVDQA

**Comment by gacholio** [comment]
Like the other platforms, we'll let the JIT do the feature check on startup and set the appropriate runtime flags.

**Comment by dnakamura** [comment]
Do we plan on using the extra registers from avx-512 (eg zmm16-31)?

**Comment by tajila** [comment]
I think the intention is that with Vector API we will use all available vector registers. @gita-omr can confirm.

**Comment by dnakamura** [comment]
I would assume so too, but I thought I would check as saving/restoring 2KB worth of registers feels a bit excessive

**Comment by gita-omr** [comment]
Yes, I afraid we need to save all of them and model it after Power where we don’t save on helpers that correspond to calls in the Java code and therefore saved by JIT already.

**Comment by dnakamura** [comment]
Have a current draft PR (#13227).

**Comment by BradleyWood** [comment]
I have a few questions / thoughts about this work (#13227)
 - Have you thought about 32-bit? We only have xmm/ymm/zmm 0-8 on 32-bit
 - What if the target has AVX2, but not AVX512? then have xmm/ymm 0-16.
 - Was this ever tested on AVX512 capable hardware?
 - Could you document your M4 macros (such as **_forloop_**)
 - What about the vector mask registers k1-k7 (ignore k0)

**Comment by dnakamura** [comment]
> I have a few questions / thoughts about this work (#13227)
> 
> * Have you thought about 32-bit? We only have xmm/ymm/zmm 0-7 on 32-bit

Yes. In the current form the 32 bit code only saves/restores [z|y|z]mm0-7

> * What if the target has AVX2, but not AVX512? then have xmm/ymm 0-15.

AVX2 = `J9_EXTENDED_RUNTIME_USE_VECTOR_REGISTERS`, AVX-512 = `J9_EXTENDED_RUNTIME_USE_EXTENDED_VECTOR_REGISTERS`, normal AVX is assumed

> * Was this ever tested on AVX512 capable hardware?

Real hardware? no. However it seems to run ok in an emulator (although it doesn't really say much since we aren't using the zmm registers)

> * Could you document your M4 macros (such as **_forloop_**)
> * What about the vector mask registers k1-k7 (ignore k0)

I forgot about the mask registers. That gets a bit tricky since we need to also know if we have AVX-512BW to know  the size, and tacking on another feature flag seems a bit excessive

**Comment by BradleyWood** [comment]
For the mask registers, I don't see a way around adding another flag. With AVX-512BW you can use KMOVQ to store all 64-bits, otherwise, with AVX512F, use KMOVW to store 16 bits.

**Comment by gacholio** [comment]
A possible option is to simply not support vector on older hardware.

**Comment by gacholio** [comment]
@vijaysun-omr @0xdaryl @tajila How new is the AVX-512BW support? Would it be reasonable to support only plain XMM and the latest greatest?

**Comment by BradleyWood** [comment]
I think most AVX-512 capable CPUs do support AVX-512BW. I can see that Knights Landing & Knights Mill do not.

**Comment by dnakamura** [comment]
also given that we have 2 bits we could map the requred 4 states, although its a bit less clean. Ie if `J9_EXTENDED_RUNTIME_USE_EXTENDED_VECTOR_REGISTERS` is set, `J9_EXTENDED_RUNTIME_USE_VECTOR_REGISTERS` could be used to indicate AVX-512BW

**Comment by BradleyWood** [comment]
@dnakamura I would say MXCSR register does not need to be preserved.

**Comment by gacholio** [comment]
The current work has been merged - closing.
